src/Features/Core/Portable/Formatting/FormattingDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,101 @@
+            => DiagnosticAnalyzerCategory.SyntaxAnalysis;
+
+        public override bool OpenFileOnly(Workspace workspace)
+            => false;
+
+        protected override void InitializeWorker(AnalysisContext context)
+            => context.RegisterSyntaxTreeAction(AnalyzeSyntaxTree);
+
+        private void AnalyzeSyntaxTree(SyntaxTreeAnalysisContext context)
+        {
+            if (!(context.Options is WorkspaceAnalyzerOptions workspaceAnalyzerOptions))
+            {
+                return;
+            }
+
+            var options = context.Options.GetDocumentOptionSetAsync(context.Tree, context.CancellationToken).GetAwaiter().GetResult();
+            if (options == null)
+            {
+                return;
+            }
+
+            var workspace = workspaceAnalyzerOptions.Services.Workspace;
+            var formattingChanges = Formatter.GetFormattedTextChanges(context.Tree.GetRoot(context.CancellationToken), workspace, options, context.CancellationToken);
+            foreach (var formattingChange in formattingChanges)
+            {
+                var change = formattingChange;
+                if (change.NewText.Length > 0 && !change.Span.IsEmpty)
+                {
+                    var oldText = context.Tree.GetText(context.CancellationToken);
+
+                    // Handle cases where the change is a substring removal from the beginning
+                    var offset = change.Span.Length - change.NewText.Length;
+                    if (offset >= 0 && oldText.GetSubText(new TextSpan(change.Span.Start + offset, change.NewText.Length)).ContentEquals(SourceText.From(change.NewText)))
+                    {
+                        change = new TextChange(new TextSpan(change.Span.Start, offset), "");
+                    }
+                    else
+                    {
+                        // Handle cases where the change is a substring removal from the end
+                        if (change.NewText.Length < change.Span.Length
+                            && oldText.GetSubText(new TextSpan(change.Span.Start, change.NewText.Length)).ContentEquals(SourceText.From(change.NewText)))
+                        {
+                            change = new TextChange(new TextSpan(change.Span.Start + change.NewText.Length, change.Span.Length - change.NewText.Length), "");
+                        }
+                    }
+                }
+
+                if (change.NewText.Length == 0 && change.Span.IsEmpty)
+                {
+                    // No actual change
+                    continue;
+                }
+
+                ImmutableDictionary<string, string> properties;
+                if (change.NewText.Length == 0)
+                {
+                    properties = RemoveTextProperties;
+                }
+                else
+                {
+                    properties = ImmutableDictionary.Create<string, string>().Add(ReplaceTextKey, change.NewText);
+                }
```

Comment:
nit: consider using ternary expressoin.

Reply:
so it reduces span to report squiggles, but it doesn't actually create multiple squiggles on the same change from formatter. in that case, is there any user experience difference by saving this change in the property bag and then apply only this at the fixer? 

you can just call formatter one the line diagnostics is on, and apply text changes the diagnostic span is contained. that way you can skip saving this info in the property bag for every diagnostic ever reported.